Source/RequestInterceptor.swift

```diff
@@ -24,6 +24,18 @@
 
 import Foundation
 
+/// A type that can inspect and optionally adapt a `URLRequest` in some manner if necessary.
+public protocol RequestAdapter {
+    /// Inspects and adapts the specified `URLRequest` in some manner and calls the completion handler with the Result.
+    ///
+    /// - Parameters:
+    ///   - urlRequest: The `URLRequest` to adapt.
+    ///   - completion: The completion handler that must be called when adaptation is complete.
+    func adapt(_ urlRequest: URLRequest, completion: @escaping (_ result: Result<URLRequest>) -> Void)
```

Comment:
Things I wondered while integrating this branch into my code and updating my adaptor unit tests...

Is there any expectation that the completion should be called asynchronously, or can it be called inline? 
If called asynchronously, is there a preferred queue? Should the adapter be passed a queue on which the completion should be called?

Reply:
No, it's specially designed for the queue not to matter. Alamofire immediately enqueues its work back onto the internal `rootQueue`.

Reply:
Maybe the docstrings should note this?

Reply:
Sure. How about "The completion handler that must be called when adaptation is complete. It may be called from any queue."